.github/workflows/main.yml

```diff
@@ -11,13 +11,35 @@ jobs:
     runs-on: ubuntu-latest
     steps:
       - uses: actions/checkout@v2
-      - name: Cache cargo
+
+      - name: Detect the installed Rust version
+        id: rustc
+        run: echo "::set-output name=version::$(rustc -V)"
+
+      - name: Cache Cargo's registry cache
+        uses: actions/cache@v2
+        with:
+          path: ~/.cargo/registry/cache
+          key: cargo-registry-cache-${{ hashFiles('**/Cargo.lock') }}
+          restore-keys: |
+            cargo-registry-cache-
+
+      - name: Cache Cargo's registry index
+        uses: actions/cache@v2
+        with:
+          path: ~/.cargo/registry/index
+          key: cargo-registry-index-${{ hashFiles('**/Cargo.lock') }}
+          restore-keys: |
+            cargo-registry-index-
+
+      - name: Cache Cargo's target directory
         uses: actions/cache@v2
         with:
           path: target
-          key: cargo-build-${{ hashFiles('**/Cargo.lock') }}
+          key: cargo-build-${{ steps.rustc.outputs.version }}-${{ hashFiles('**/Cargo.lock') }}
           restore-keys: |
-            cargo-build-
+            cargo-build-${{ steps.rustc.outputs.version }}-
+
```

Comment:
What does not including `Cargo.lock` in the restore key get you?

Reply:
GitHub Actions's caching works roughly this way when it tries to restore something:

* If a cache with the exact key is present (in this case, when both the Rust version and the `Cargo.lock` are the same) it will download that specific cache.
* If an exact cache match is not present, it will try downloading the latest cache matching the prefix in the `restore-keys`.

So, what this snippet does is it tries to download the cached target directory for the right combination of Rust version and `Cargo.lock`, and if that fails it will try downloading the latest cached target directory for that Rust version, regardless of the contents of the `Cargo.lock`